Blazor Hybrid dev tools topic

guardrex opened the pull request:
Fixes #25382
Addresses #24956

[Internal Review Topic](https://review.docs.microsoft.com/en-us/aspnet/core/blazor/hybrid/developer-tools?view=aspnetcore-6.0&branch=pr-en-us-25384)

Notes:

* The PR currently uses `BlazorWebView`, and I'll change that if it makes more sense to use `WebView`.
* Please confirm/update the UID (for cross-linking) and the filename (for the topic's URL). These are the most important aspects to get right for the future.

# 🇺🇦

TanayParikh commented on aspnetcore/toc.yml:
```diff
@@ -351,6 +351,8 @@ items:
                     uid: blazor/hybrid/tutorials/wpf
               - name: Routing and navigation
                 uid: blazor/hybrid/routing
+              - name: Developer tools
+                uid: blazor/hybrid/developer-tools
```
```suggestion
              - name: Accessing Browser Developer tools
                uid: blazor/hybrid/accessing-browser-developer-tools
```

guardrex replied:
Do we want to make it this specific? I mean this scopes down to just "accessing" and not "using" or doing anything else. I generally try to keep these generic and broad so that we can place any kind of "developer tools" guidance.

Another possibility is to make the whole topic **_even more generic_** by making this guidance part of a *Troubleshoot*-type of topic or an *Error handling*-type of topic.

TanayParikh replied:
I like `error handling`. I'm a bit hesitant to do dev tools as users may think this is guidance for vs/vscode which it is not (yet?). There are a few other things which may be upcoming on the error handling front so this would be a good section for that as well.

guardrex replied:
Ok ... 👂 good to me. I'll pitch a revision for it on Thursday morning with the additional updates. I'll ping u back tomorrow.

TanayParikh commented on aspnetcore/blazor/hybrid/developer-tools.md:
```diff
@@ -0,0 +1,54 @@
+ms.author: riande
+ms.custom: "mvc"
+ms.date: 03/23/2022
+no-loc: ["Blazor Hybrid", Home, Privacy, Kestrel, appsettings.json, "ASP.NET Core Identity", cookie, Cookie, Blazor, "Blazor Server", "Blazor WebAssembly", "Identity", "Let's Encrypt", Razor, SignalR]
```
Are these meta keywords? Can we add `.NET MAUI`?

guardrex replied:
MT doesn't run on Blazor docs node. It's all HT ("human translation") here. However, it's possible that some or all of the topics would become MT at some point. Yes, I think it's a good idea to add it. I'll do that, but I'll do it on a separate PR. I'll note it on my tracking issue so that I don't lose this.

guardrex replied:
Oh, sorry ... I missed the ❓ ...

> Are these meta keywords?

No, they're words/phrases to avoid for localization. .NET MAUI doesn't translate, but it would be safe to add it to the list, and I've made a note to do so.

TanayParikh replied:
Let's add `Mac Catalyst` as well, I've seen that unintentionally localized in the past.

guardrex replied:
Got it ... I added "Mac Catalyst" to the UE tracking issue. It won't get lost.

TanayParikh commented on aspnetcore/blazor/hybrid/developer-tools.md:
```diff
@@ -0,0 +1,54 @@
+ms.custom: "mvc"
+ms.date: 03/23/2022
+no-loc: ["Blazor Hybrid", Home, Privacy, Kestrel, appsettings.json, "ASP.NET Core Identity", cookie, Cookie, Blazor, "Blazor Server", "Blazor WebAssembly", "Identity", "Let's Encrypt", Razor, SignalR]
+uid: blazor/hybrid/developer-tools
```
```suggestion
uid: blazor/hybrid/accessing-browser-developer-tools
```

TanayParikh commented on aspnetcore/blazor/hybrid/developer-tools.md:
```diff
@@ -0,0 +1,54 @@
+
+To use developer tools with a .NET MAUI Blazor app, use the guidance in the following sections depending on the operating system used to run the app:
+
+* [iOS](#ios)
+* [Android](#android)
+
```
@guardrex this wasn't covered as part of the original issue, but would you be able to add additional guidance for Windows & MacOS?

- Windows should just be a matter of CTRL+SHIFT+I within the BlazorWebView to open the dev tools.
- Mac should be an essentially identical process to ios.

guardrex replied:
Yes, I'll do that first thing Thursday morning and ping u back.

TanayParikh commented on aspnetcore/blazor/hybrid/developer-tools.md:
```diff
@@ -0,0 +1,54 @@
+
+1. Open desktop Safari.
+1. Select the **Preferences** > **Advanced** > **Show Develop** menu item.
+1. Run the .NET MAUI Blazor app in the Visual Studio emulator.
```
> Visual Studio emulator

Not sure what this may be referring to? Perhaps:

```suggestion
1. Run the .NET MAUI Blazor app in the iOS simulator.
```

TanayParikh commented on aspnetcore/blazor/hybrid/developer-tools.md:
```diff
@@ -0,0 +1,54 @@
+1. Open desktop Safari.
+1. Select the **Preferences** > **Advanced** > **Show Develop** menu item.
+1. Run the .NET MAUI Blazor app in the Visual Studio emulator.
+1. Return to desktop Safari. Select **Develop** > **Simulator** > **0.0.0.0**. If multiple entries for **0.0.0.0** are present, select the entry that highlights the `BlazorWebView`. The `BlazorWebView` is highlighted in blue in the Visual Studio emulator when the correct **0.0.0.0** entry is selected.
```
```suggestion
1. Return to desktop Safari. Select **Develop** > **Simulator** > **0.0.0.0**. If multiple entries for **0.0.0.0** are present, select the entry that highlights the `BlazorWebView`. The `BlazorWebView` is highlighted in blue in the iOS simulator when the correct **0.0.0.0** entry is selected.
```

TanayParikh commented on aspnetcore/blazor/hybrid/developer-tools.md:
```diff
@@ -0,0 +1,54 @@
+
+   ![Safari Develop Simulator open showing two entries for "0.0.0.0" with the second entry selected because it highlights the BlazorWebView in the Visual Studio emulator UI.](~/blazor/hybrid/developer-tools/_static/ios.png)
+
+1. The Simulator window opens, and the **Web Inspector** appears for the `BlazorWebView`.
```
The simulator should already be up and running at this time.

```suggestion
1. The **Web Inspector** window appears for the `BlazorWebView`.
```

TanayParikh commented on aspnetcore/blazor/hybrid/developer-tools.md:
```diff
@@ -0,0 +1,54 @@
+   ![Safari Develop Simulator open showing two entries for "0.0.0.0" with the second entry selected because it highlights the BlazorWebView in the Visual Studio emulator UI.](~/blazor/hybrid/developer-tools/_static/ios.png)
+
+1. The Simulator window opens, and the **Web Inspector** appears for the `BlazorWebView`.
+1. Inspect the **Console** tab to see the console messages, which includes any exception messages generated by the framework or developer code:
```
I wonder if the current verbiage here may sound restrictive? Inspecting using the console tab is _one of_ the things you can do. The current phrasing sounds "instructive" as if it is the first/only thing you can do? Maybe I'm reading too much into it?

TanayParikh commented on aspnetcore/blazor/hybrid/developer-tools.md:
```diff
@@ -0,0 +1,54 @@
+
+Use the following steps to use browser developer tools with an Android app. The following example shows how to inspect an unhandled exception in the developer tools console log:
+
+1. Start the Visual Studio Android emulator and navigate to an app page that uses a `BlazorWebView`.
```
```suggestion
1. Start the Android emulator and navigate to an app page that uses a `BlazorWebView`.
```

TanayParikh commented on aspnetcore/blazor/hybrid/developer-tools.md:
```diff
@@ -0,0 +1,54 @@
+
+   ![Microsoft Edge Devices showing the BlazorWebView's "inspect" link button to open developer tools.](~/blazor/hybrid/developer-tools/_static/android.png)
+
+1. Inspect the **Console** tab to see the console messages, which includes any exception messages generated by the framework or developer code:
```
See above comment.

TanayParikh reviewed (commented):
Thanks Luke, this is a great start! I've left some clarifying comments below.

Would it be possible to add instructions for Windows/MacOS as well, they should just be an extension of the existing iOS instructions (for Mac) and CTRL+SHIFT+I (for Windows)

guardrex commented:
@TanayParikh Updates made.

For the Safari image for macOS, I had to Photoshop your iOS image down to just the Web Inspector, which might be ok. The problem is that my personal Mac (that I own and have been using for the last several years) is now too old to run .NET. I can't get one from Aquent 😠. There's nothing that I can do about that. Dan and Artak are aware of the situation. Anyway ... if you want a macOS-specific Safari image (i.e., that shows the app running, too), you'd need to create the image for me and post it in an issue comment here.

Also, I added a "doc pivot" (tab control) for the OSs. We usually do that for OS guidance. If you prefer that I split it back out into sections, that's fine ... I can do that. Let me know. 👂

TanayParikh commented on aspnetcore/zone-pivot-groups.yml:
```diff
@@ -18,3 +18,15 @@ groups:
       title: Blazor Server
     - id: webassembly
       title: Blazor WebAssembly
+- id: blazor-hybrid-operating-systems
+  title: Operating System
+  prompt: Choose an operating system
```
How about "Target Platform" or "Target Operating System" to distinguish between the host development machine operating system (Windows/Mac) vs the "target" (Windows/Mac/iOS/Android).

guardrex replied:
```suggestion
  prompt: Target operating system
```

TanayParikh commented on aspnetcore/blazor/hybrid/handle-errors.md:
```diff
@@ -0,0 +1,82 @@
+
+1. Open desktop Safari.
+1. Select the **Preferences** > **Advanced** > **Show Develop** menu item.
+1. Run the .NET MAUI Blazor app in the iOS simulator.
```
```suggestion
1. Run the .NET MAUI Blazor app in the iOS simulator and navigate to an app page that uses a `BlazorWebView`.
```

TanayParikh commented on aspnetcore/blazor/hybrid/handle-errors.md:
```diff
@@ -0,0 +1,82 @@
+
+:::zone pivot="macos"
+
+To use Safari developer tools with a macOS app:
+
+1. Open desktop Safari.
+1. Select the **Preferences** > **Advanced** > **Show Develop** menu item.
```
Was just trying this out and I was mistaken there is a bit more nuance than just the iOS steps. Looking into this and will update...

TanayParikh replied:
Looks like this may be a regression on Mac/Safari: https://apple.stackexchange.com/questions/416277/how-to-inspect-safari-web-inspector-aka-devtools-on-devtools

Looking for workarounds...

TanayParikh replied:
Okay this looks like a genuine bug. Can we just remove the macos section entirely for now?

Bug: https://github.com/dotnet/maui/issues/5561

TanayParikh commented on aspnetcore/blazor/hybrid/handle-errors.md:
```diff
@@ -0,0 +1,82 @@
+
+To use browser developer tools with a Windows app:
+
+1. Start the app with the **Windows Machine** Start button.
+1. Navigate to a page that uses the `BlazorWebView`.
```
```suggestion
1. Run the .NET MAUI Blazor app for Windows and navigate to an app page that uses a `BlazorWebView`.
```

To keep consistent with other platforms, I don't think we need to discuss the means by which to start the application here.

guardrex commented:
Ok ... updates performed thus far.

btw ... wrt ...

> you'd need to create the image for me and post it in an issue comment here.

If you do produce an image for Safari with macOS ...

It's not critical, but I recommend naming the test app "MAUIBlazor" to match the general Blazor Hybrid app naming that docs will adopt. This is a change that I'll be making shortly to the tutorial for .NET MAUI. I have the images. I just haven't had a chance to swap them in yet ... maybe on Friday.

guardrex commented:
@TanayParikh ... I commented-out the macOS piece for that pivot, tell the reader to use the iOS instructions, and link to reload the article to that pivot.

TanayParikh commented on aspnetcore/blazor/hybrid/handle-errors.md:
```diff
@@ -55,6 +55,9 @@ To use Safari developer tools with an iOS app:
 
 :::zone pivot="macos"
 
+To use Safari developer tools with a macOS app, use the instructions for [iOS](?pivots=ios).
```
```suggestion
Using browser developer tools with Mac Catalyst applications isn't currently supported. Please use the guidance for [iOS](?pivots=ios) applications as the behavior should be very similar between the two platforms..
```

TanayParikh commented on aspnetcore/blazor/hybrid/handle-errors.md:
```diff
@@ -0,0 +1,94 @@
+
+## Browser developer tools with .NET MAUI Blazor
+
+Ensure your current Blazor Hybrid project is configured to support browser developer tools. Add the following to your `CreateMauiApp` method. The `CreateMauiApp` method should be within your main `Program` or `Startup` file and contain `services.AddMauiBlazorWebView()`.
```
```suggestion
Ensure your current Blazor Hybrid project is configured to support browser developer tools. Add the following to your `CreateMauiApp` method if it is not already present. The `CreateMauiApp` method should be within your main `Program` or `Startup` file and contain `services.AddMauiBlazorWebView()`.
```

TanayParikh replied:
```suggestion
Ensure your current Blazor Hybrid project is configured to support browser developer tools. Add the following to your `CreateMauiApp` method. The `CreateMauiApp` method is likely within your main `Program` or `Startup` file and contain `services.AddMauiBlazorWebView()`.
```

TanayParikh commented on aspnetcore/blazor/hybrid/handle-errors.md:
```diff
@@ -0,0 +1,94 @@
+
+## Browser developer tools with .NET MAUI Blazor
+
+Ensure your current Blazor Hybrid project is configured to support browser developer tools. Add the following to your `CreateMauiApp` method if it is not already present. The `CreateMauiApp` method should be within your main `Program` or `Startup` file and contain `services.AddMauiBlazorWebView()`.
```
```suggestion
Ensure your current Blazor Hybrid project is configured to support browser developer tools. Add the following to your `CreateMauiApp` method if it is not already present. The `CreateMauiApp` method is likely within your main `Program` or `Startup` file and contain `services.AddMauiBlazorWebView()`.
```

TanayParikh commented on aspnetcore/blazor/hybrid/handle-errors.md:
```diff
@@ -0,0 +1,94 @@
+
+## Browser developer tools with .NET MAUI Blazor
+
+Ensure your current Blazor Hybrid project is configured to support browser developer tools. Add the following to your `CreateMauiApp` method if it is not already present. The `CreateMauiApp` method is likely within your main `Program` or `Startup` file and contain `services.AddMauiBlazorWebView()`.
```
```suggestion
Ensure your current Blazor Hybrid project is configured to support browser developer tools. Add the following to your `CreateMauiApp` method if it is not already present. The `CreateMauiApp` method is likely within your main `Program` or `Startup` file and contains `services.AddMauiBlazorWebView()`.
```

TanayParikh reviewed (approved):
Thanks, this looks great, lets ship it!

guardrex commented:
Does Dan want to see it first?

guardrex commented:
@danroth27 ... Do you want to review/approve this before we merge?

TanayParikh commented on aspnetcore/blazor/hybrid/handle-errors.md:
````diff
@@ -0,0 +1,94 @@
+#if DEBUG
+    services.AddBlazorWebViewDeveloperTools();
+#endif
+```
+
+:::zone pivot="windows"
````
```suggestion
```

> [!NOTE]
> Guidance on popular browsers' developer tools can be found in the documentation of each browser maintainer:
>
> * [Chrome DevTools](https://developer.chrome.com/docs/devtools/)
> * [Microsoft Edge Developer Tools overview](/microsoft-edge/devtools-guide-chromium/)
> * [Safari Web Inspector](https://support.apple.com/guide/safari-developer/welcome/mac)

:::zone pivot="windows"
```

TanayParikh replied:
```suggestion

> [!NOTE]
> Guidance on popular browsers' developer tools can be found in the documentation of each browser maintainer:
>
> * [Chrome DevTools](https://developer.chrome.com/docs/devtools/)
> * [Microsoft Edge Developer Tools overview](/microsoft-edge/devtools-guide-chromium/)
> * [Safari Web Inspector](https://support.apple.com/guide/safari-developer/welcome/mac)

:::zone pivot="windows"
```

guardrex commented:
Closed and re-opened because it looked like the build was hung.

danroth27 reviewed (commented):
Looks great! My one piece of feedback is the title. When I was looking for this doc in the TOC I scanned for "browser developer tools" but I didn't see it. I thought "Handle errors" meant runtime error management.
